toolchain/lower/file_context.cpp

```diff
@@ -116,6 +121,166 @@ auto FileContext::Run() -> std::unique_ptr<llvm::Module> {
   return std::move(llvm_module_);
 }
 
+auto FileContext::CoalesceEquivalentSpecifics() -> void {
+  // TODO: non-determinism check: does it matter the order of processing
+  // generic_ids? Non-issue for a given generic_id, as the order of the
+  // specifics for a generic is fixed.
+  lowered_specifics.ForEach([&](auto /* generic_id */, auto& specifics) {
+    for (unsigned i = 0; i < specifics.size() - 1; ++i) {
+      if (replaced_specifics.Contains(specifics[i])) {
+        specifics[i] = specifics[specifics.size() - 1];
+        specifics.pop_back();
+        --i;
+        continue;
+      }
+      for (unsigned j = i + 1; j < specifics.size(); ++j) {
```

Comment:
It looks like this is going to be at least O(N^2) when there are N specifics. We've generally tried really hard to avoid quadratic running times, so that seems like it might be a dealbreaker here, but I don't understand the problem domain well enough to suggest alternatives.

Reply:
To note, we'd discussed this, but this may still be worth a TODO to help us not forget.